ESQL: GROUP BY ALL with the dimensions output

leontyevdv opened the pull request:
The second part of the https://github.com/elastic/elasticsearch/issues/136253. 

Load a list of dimensions and output them in the _timeseries column instead of _tsid ([here is a doc about it](https://docs.google.com/document/d/1zyHNj6II0ysp_DwM1ilt7K5XxhlcJgqp4GmkTabx-hk/edit?tab=t.0)).

The first part is here: https://github.com/elastic/elasticsearch/pull/137367

Part of https://github.com/elastic/elasticsearch/issues/136253

not-napoleon committed:
add capability and initial CSV test (WIP)

not-napoleon committed:
tests (that don't pass)

not-napoleon committed:
Merge branch 'main' into esql-group-by-all-over-time

not-napoleon committed:
tests and tracing notes

not-napoleon committed:
Merge branch 'main' into esql-group-by-all-over-time

not-napoleon committed:
added rate agg to tests

not-napoleon committed:
add test gates

not-napoleon committed:
detect when group by all should kick in

not-napoleon committed:
add outline for the rest of the steps

not-napoleon committed:
why does this throw?

not-napoleon committed:
load dimension and metric data in the tests

not-napoleon committed:
build the values aggs for the dimensions

not-napoleon committed:
work around output changed verification

not-napoleon committed:
notes for future me, and others

not-napoleon committed:
draft of analyzer rule

not-napoleon committed:
wire up the rule

not-napoleon committed:
many tests passing

not-napoleon committed:
Merge branch 'main' into esql-group-by-all-over-time

 Conflicts:
	x-pack/plugin/esql/qa/testFixtures/src/main/resources/k8s-timeseries.csv-spec

not-napoleon committed:
failing for a new reason

Commit:
[CI] Auto commit changes from spotless

not-napoleon committed:
misc cleanups

not-napoleon committed:
Merge branch 'main' into esql-group-by-all-over-time

 Conflicts:
	x-pack/plugin/esql/qa/testFixtures/src/main/java/org/elasticsearch/xpack/esql/LoadMapping.java
	x-pack/plugin/esql/qa/testFixtures/src/main/resources/k8s-timeseries.csv-spec

not-napoleon committed:
more testing

not-napoleon committed:
Merge remote-tracking branch 'refs/remotes/not-napoleon/esql-group-by-all-over-time' into esql-group-by-all-over-time

not-napoleon committed:
forgot to add this test yesterday

Commit:
[CI] Auto commit changes from spotless

Commit:
[CI] Update transport version definitions

leontyevdv committed:
Merge branch 'main' into esql-group-by-all-over-time

# Conflicts:
#	server/src/main/resources/transport/upper_bounds/8.18.csv
#	server/src/main/resources/transport/upper_bounds/8.19.csv
#	server/src/main/resources/transport/upper_bounds/9.0.csv
#	server/src/main/resources/transport/upper_bounds/9.1.csv
#	server/src/main/resources/transport/upper_bounds/9.2.csv
#	server/src/main/resources/transport/upper_bounds/9.3.csv
#	x-pack/plugin/esql/qa/testFixtures/src/main/resources/k8s-timeseries.csv-spec
#	x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/logical/TranslateTimeSeriesAggregate.java
#	x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/logical/Drop.java

leontyevdv committed:
Merge branch 'main' into esql-group-by-all-over-time

leontyevdv committed:
Fix unit tests

leontyevdv committed:
Merge branch 'main' into feature/esql-group-by-all

leontyevdv committed:
ESQL: Add GROUP BY ALL

Group by _tsid and add it to the output.

Closes #136253

leontyevdv committed:
ESQL: Add GROUP BY ALL

Group by _tsid and add it to the output.

Closes #136253

leontyevdv committed:
Merge branch 'main' into feature/esql-group-by-all

leontyevdv committed:
Merge branch 'main' into feature/esql-group-by-all

leontyevdv committed:
ESQL: Add GROUP BY ALL

Group by _tsid and add it to the output in the _timeseries metadata
attribute.

Closes #136253

leontyevdv committed:
Merge branch 'main' into feature/esql-group-by-all

leontyevdv committed:
Merge branch 'main' into feature/esql-group-by-all

leontyevdv committed:
ESQL: Add GROUP BY ALL

Group by _tsid and add it to the output in the _timeseries metadata
attribute.

Part of #136253

leontyevdv committed:
Merge branch 'main' into feature/esql-group-by-all

# Conflicts:
#	x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/logical/TranslateTimeSeriesAggregate.java

leontyevdv committed:
ESQL: Add GROUP BY ALL

Group by _tsid and add it to the output in the _timeseries metadata
attribute.

Part of #136253

leontyevdv committed:
ESQL: Add GROUP BY ALL

Fix tests

Part of #136253

leontyevdv committed:
Update docs/changelog/137367.yaml

leontyevdv committed:
ESQL: Add GROUP BY ALL

Fix tests

Part of #136253

leontyevdv committed:
Merge branch 'main' into feature/esql-group-by-all

leontyevdv committed:
ESQL: Add GROUP BY ALL

Cleanup comments

Part of #136253

leontyevdv committed:
Merge remote-tracking branch 'origin/feature/esql-group-by-all' into feature/esql-group-by-all

leontyevdv committed:
ESQL: Add GROUP BY ALL

Cleanup

Part of #136253

leontyevdv committed:
ESQL: Add GROUP BY ALL

Cleanup

Part of #136253

leontyevdv committed:
ESQL: Add GROUP BY ALL

Cleanup

Part of #136253

leontyevdv committed:
ESQL: Add GROUP BY ALL

Ignore GroupByAll in GenerativeForkRestTest

Part of #136253

leontyevdv committed:
Update docs/changelog/137367.yaml

leontyevdv committed:
ESQL: Add GROUP BY ALL

Remove Drop JavaDoc

Part of #136253

leontyevdv committed:
ESQL: Add GROUP BY ALL

Address the PR's comments

Part of #136253

leontyevdv committed:
ESQL: Add GROUP BY ALL

Address the PR's comments

Part of #136253

leontyevdv committed:
Merge branch 'main' into feature/esql-group-by-all

# Conflicts:
#	x-pack/plugin/esql/qa/testFixtures/src/main/resources/k8s-timeseries.csv-spec

leontyevdv committed:
ESQL: Add GROUP BY ALL

Address the PR's comments

Part of #136253

leontyevdv committed:
ESQL: Add GROUP BY ALL

Address the PR's comments

Part of #136253

leontyevdv committed:
ESQL: Add GROUP BY ALL

Address the PR's comments

Part of #136253

leontyevdv committed:
Merge branch 'main' into feature/esql-group-by-all

# Conflicts:
#	x-pack/plugin/esql/qa/server/src/main/java/org/elasticsearch/xpack/esql/qa/rest/generative/GenerativeForkRestTest.java

leontyevdv committed:
Merge branch 'main' into feature/esql-group-by-all

leontyevdv committed:
ESQL: Add GROUP BY ALL

Address the PR's comments

Part of #136253

leontyevdv committed:
ESQL: Add GROUP BY ALL

Remove support of the grouping attributes for the bare time-series
aggregates.

Part of #136253

leontyevdv committed:
Merge branch 'main' into feature/esql-group-by-all

leontyevdv committed:
ESQL: Add GROUP BY ALL

Fix merge.

Part of #136253

leontyevdv committed:
ESQL: Add GROUP BY ALL

Add output of the dimension list into the _timeseries column.

Part of #136253

leontyevdv committed:
Update docs/changelog/138595.yaml

leontyevdv committed:
ESQL: Add GROUP BY ALL

Add output of the dimension list into the _timeseries column.

Part of #136253

leontyevdv committed:
Merge branch 'main' into feature/esql-group-by-all-dimensions

# Conflicts:
#	x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/physical/local/ExtractDimensionFieldsAfterAggregation.java

Commit:
[CI] Auto commit changes from spotless

leontyevdv committed:
ESQL: Add GROUP BY ALL

Add output of the dimension list into the _timeseries column.

Part of #136253

leontyevdv committed:
Merge remote-tracking branch 'origin/feature/esql-group-by-all-dimensions' into feature/esql-group-by-all-dimensions

leontyevdv committed:
Merge branch 'main' into feature/esql-group-by-all-dimensions

leontyevdv committed:
Merge branch 'main' into feature/esql-group-by-all-dimensions

dnhatn commented:
I discussed this with Dima, and Dima will explore an alternative approach where we fuse the dimension fields into the _source without introducing the _timeseries mapper.

leontyevdv committed:
ESQL: Add GROUP BY ALL

Add output of the dimension list into the _timeseries column.

Part of #136253

leontyevdv committed:
ESQL: Add GROUP BY ALL

Add output of the dimension list into the _timeseries column.

Part of #136253

leontyevdv committed:
ESQL: Add GROUP BY ALL

Add output of the dimension list into the _timeseries column.

Part of #136253

leontyevdv committed:
ESQL: Add GROUP BY ALL

Add output of the dimension list into the _timeseries column.

Part of #136253

leontyevdv committed:
ESQL: Add GROUP BY ALL

Add output of the dimension list into the _timeseries column.

Part of #136253

leontyevdv committed:
ESQL: Add GROUP BY ALL

Add output of the dimension list into the _timeseries column.

Part of #136253

leontyevdv committed:
Merge branch 'main' into feature/esql-group-by-all-dimensions

kkrik-es reviewed (commented):
Looks good!

leontyevdv committed:
ESQL: Add GROUP BY ALL

Add output of the dimension list into the _timeseries column.

Part of #136253

leontyevdv committed:
Merge branch 'main' into feature/esql-group-by-all-dimensions

dnhatn reviewed (commented):
One comment, but this looks good. Thanks Dima!

leontyevdv committed:
Merge branch 'main' into feature/esql-group-by-all-dimensions

leontyevdv committed:
ESQL: Add GROUP BY ALL

Add output of the dimension list into the _timeseries column.

Part of #136253

leontyevdv committed:
ESQL: Add GROUP BY ALL

Add output of the dimension list into the _timeseries column.

Part of #136253

leontyevdv committed:
Merge branch 'main' into feature/esql-group-by-all-dimensions

leontyevdv committed:
ESQL: Add GROUP BY ALL

Fix a comment

Part of #136253

leontyevdv committed:
Merge remote-tracking branch 'origin/feature/esql-group-by-all-dimensions' into feature/esql-group-by-all-dimensions

kkrik-es reviewed (commented):
Well done Dima. Will leave it to Nhat to stamp.

leontyevdv committed:
Merge branch 'main' into feature/esql-group-by-all-dimensions

dnhatn reviewed (approved):
One comment, but this looks great. Thanks Dima!

leontyevdv committed:
Merge branch 'main' into feature/esql-group-by-all-dimensions

leontyevdv committed:
ESQL: Add GROUP BY ALL

Move the dimensions extraction into the TimeSeriesMetadataFieldBlockLoader

Part of #136253

leontyevdv committed:
Merge remote-tracking branch 'origin/feature/esql-group-by-all-dimensions' into feature/esql-group-by-all-dimensions

leontyevdv committed:
ESQL: Add GROUP BY ALL

Add tests

Part of #136253

leontyevdv committed:
ESQL: Add GROUP BY ALL

Add tests

Part of #136253

dnhatn reviewed (approved):
Thanks Dima for another iteration!

leontyevdv committed:
ESQL: Add GROUP BY ALL

Polish code

Part of #136253

kkrik-es commented on server/src/main/java/org/elasticsearch/index/mapper/TimeSeriesMetadataFieldBlockLoader.java:
```diff
@@ -0,0 +1,99 @@
+    private static class TimeSeries extends BlockStoredFieldsReader {
+        @Override
+        public void read(int docId, StoredFields storedFields, Builder builder) throws IOException {
+            // TODO support appending BytesReference
```
You'd probably need to have a time-series index with hardcoded routing path that includes no dimension fields. This is not very interesting, admittedly, but can happen I guess. A graceful failure is fine, so long as we don't produce a cryptic error.
